feature/s3/transfermanager/api_op_PutObject.go

```diff
@@ -0,0 +1,995 @@
+	for u.geterr() == nil && err == nil {
+		partNum++
+		var (
+			data         io.Reader
+			nextChunkLen int
+			ok           bool
+		)
+		data, nextChunkLen, cleanup, err = u.nextReader(ctx)
+		ok, err = u.shouldContinue(partNum, nextChunkLen, err)
+		if !ok {
+			cleanup()
+			if err != nil {
+				u.seterr(err)
+			}
+			break
+		}
+
+		ch <- ulChunk{buf: data, partNum: aws.Int32(partNum), cleanup: cleanup}
+	}
+
+	// close the channel, wait for workers and complete upload
+	close(ch)
+	u.wg.Wait()
+	completeOut := u.complete(ctx, clientOptions...)
+
+	if err := u.geterr(); err != nil {
+		return nil, &multipartUploadError{
+			err:      err,
+			uploadID: *u.uploadID,
+		}
+	}
+
+	var out PutObjectOutput
+	out.mapFromCompleteMultipartUploadOutput(completeOut, aws.ToString(params.Bucket), aws.ToString(u.uploadID), u.parts)
+	return &out, nil
+}
+
+func (u *multiUploader) shouldContinue(part int32, nextChunkLen int, err error) (bool, error) {
+	if err != nil && err != io.EOF {
+		return false, fmt.Errorf("read multipart upload data failed, %w", err)
+	}
+
+	if nextChunkLen == 0 {
+		// No need to upload empty part, if file was empty to start
+		// with empty single part would of been created and never
+		// started multipart upload.
+		return false, nil
+	}
+
+	// This upload exceeded maximum number of supported parts, error now.
+	if part > defaultMaxUploadParts {
+		return false, fmt.Errorf(fmt.Sprintf("exceeded total allowed S3 limit MaxUploadParts (%d). Adjust PartSize to fit in this limit", defaultMaxUploadParts))
+	}
+
+	return true, err
+}
+
+// readChunk runs in worker goroutines to pull chunks off of the ch channel
+// and send() them as UploadPart requests.
+func (u *multiUploader) readChunk(ctx context.Context, ch chan ulChunk, clientOptions ...func(*s3.Options)) {
+	defer u.wg.Done()
+	for {
+		data, ok := <-ch
+
+		if !ok {
+			break
+		}
+
+		if u.geterr() == nil {
+			if err := u.send(ctx, data, clientOptions...); err != nil {
+				u.seterr(err)
+			}
+		}
+
+		data.cleanup()
+	}
+}
+
+// send performs an UploadPart request and keeps track of the completed
+// part information.
+func (u *multiUploader) send(ctx context.Context, c ulChunk, clientOptions ...func(*s3.Options)) error {
+	params := u.in.mapUploadPartInput(c.buf, c.partNum, u.uploadID)
+	resp, err := u.options.S3.UploadPart(ctx, params, clientOptions...)
+	if err != nil {
+		return err
+	}
+
+	var completed types.CompletedPart
+	completed.MapFrom(resp, c.partNum)
+
+	u.m.Lock()
+	u.parts = append(u.parts, completed)
+	u.m.Unlock()
+
+	return nil
+}
+
+// geterr is a thread-safe getter for the error object
+func (u *multiUploader) geterr() error {
+	u.m.Lock()
+	defer u.m.Unlock()
+
+	return u.err
+}
+
+// seterr is a thread-safe setter for the error object
+func (u *multiUploader) seterr(e error) {
+	u.m.Lock()
+	defer u.m.Unlock()
+
+	u.err = e
+}
+
+// fail will abort the multipart unless LeavePartsOnError is set to true.
```

Comment:
Which option is this?

I suspect we are trying to do this

> As a result, it might be necessary to abort a given multipart upload multiple times in order to completely free all storage consumed by all parts.
> To verify that all parts have been removed and prevent getting charged for the part storage, you should call the [ListParts](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListParts.html) API operation and ensure that the parts list is empty.